src/rust/runtime/memory/buffer/demibuffer.rs

```diff
@@ -0,0 +1,824 @@
+    next: Option<NonNull<MetaData>>,
+
+    // Various fields for TX offload.
+    _tx_offload: u64,
+
+    // Pointer to shared info (rte_mbuf_ext_shared_info).  DPDK uses this for external MBufs.
+    _shinfo: u64,
+
+    // Size of private data (between rte_mbuf struct and the data) in direct MBufs.
+    _priv_size: u16,
+    // Timesync flags for use with IEEE 1588 "Precision Time Protocol" (PTP).
+    _timesync: u16,
+    // Reserved for dynamic fields.
+    _dynfield: [u32; 9],
+}
+
+// Check MetaData structure alignment and size at compile time.
+// Note that alignment must be specified via a literal value in #[repr(align(64))] above, so a defined constant can't
+// be used.  So, if the alignment assert is firing, change the value in the align() to match CPU_DATA_CACHE_LINE_SIZE.
+const _: () = assert!(std::mem::align_of::<MetaData>() == arch::CPU_DATA_CACHE_LINE_SIZE);
+const _: () = assert!(std::mem::size_of::<MetaData>() == 2 * arch::CPU_DATA_CACHE_LINE_SIZE);
+
+// MetaData "offload flags".  These exactly mimic those of DPDK MBufs.
+
+// Indicates this MetaData struct doesn't have the actual data directly attached, but rather this MetaData's buf_addr
+// points to another MetaData's directly attached data.
+const METADATA_F_INDIRECT: u64 = 1 << 62;
+
+impl MetaData {
+    // Note on Reference Counts:
+    // Since we are currently single-threaded, there is no need to use atomic operations for refcnt manipulations.
+    // We should rework the implementation of inc_refcnt() and dec_refcnt() to use atomic operations if this changes.
+    // Also, we intentionally don't check for refcnt overflow.  This matches DPDK's behavior, which doesn't check for
+    // reference count overflow either (we're highly unlikely to ever have 2^16 copies of the same data).
+
+    // Increments the reference count and returns the new value.
+    #[inline]
+    pub fn inc_refcnt(&mut self) -> u16 {
+        self.refcnt += 1;
+        self.refcnt
+    }
+
+    // Decrements the reference count and returns the new value.
+    #[inline]
+    pub fn dec_refcnt(&mut self) -> u16 {
+        self.refcnt -= 1;
+        self.refcnt
+    }
+
+    // Gets the MetaData for the last segment in the buffer chain.
+    #[inline]
+    pub fn get_last_segment(&mut self) -> &mut MetaData {
+        let mut md: &mut MetaData = self;
+        while md.next.is_some() {
+            // Safety: The call to as_mut is safe, as the pointer is aligned and dereferenceable, and the MetaData
+            // struct it points to is initialized properly.
+            md = unsafe { md.next.unwrap().as_mut() };
+        }
+        &mut *md
+    }
+}
+
+// DemiBuffer type tags.
+// Since our MetaData structure is 64-byte aligned, the lower 6 bits of a pointer to it are guaranteed to be zero.
+// We currently only use the lower 2 of those bits to hold the type tag.
+#[derive(PartialEq)]
+enum Tag {
+    Heap = 1,
+    #[cfg(feature = "libdpdk")]
+    Dpdk = 2,
+}
+const TAG_MASK: usize = 0x3;
```

Comment:
Could you move this to the namespace of the tag?

Ex:

```rust
impl Tag {
    pub const MASK: usize = 0x3;
}
```

Reply:
Good idea, it's cleaner that way.

Changed and committed.